Framework/PythonInterface/plugins/algorithms/FindUBFromScatteringPlane.py

```diff
@@ -0,0 +1,135 @@
+
+    # Can change these to be 2 arrays of abc, and alpha,beta,gamma if that is preferred
+    def PyExec(self):
+        # getting lattice parameters
+        a = self.getProperty("a").value
+        b = self.getProperty("b").value
+        c = self.getProperty("c").value
+
+        alpha = self.getProperty("alpha").value
+        beta = self.getProperty("beta").value
+        gamma = self.getProperty("gamma").value
+
+        # getting vectors to define the plane
+        vector_1 = np.array(self.getProperty("Vector1").value)
+        vector_1 = V3D(vector_1[0], vector_1[1], vector_1[2])
+        vector_2 = np.array(self.getProperty("Vector2").value)
+        vector_2 = V3D(vector_2[0], vector_2[1], vector_2[2])
+
+        # Find a UB with the correct scattering plane (even if `u` and `v` passed to `SetUB` are not orthogonal)
+
+        peak_table = self.getProperty("PeaksWorkspace").value
+        self.exec_child_alg("SetUB", Workspace=peak_table, u=vector_1, v=vector_2, a=a, b=b, c=c, alpha=alpha, beta=beta, gamma=gamma)
+        lattice = peak_table.sample().getOrientedLattice()
+        ub_arb = lattice.getUB()
+
+        vertical_dir = peak_table.getInstrument().getReferenceFrame().vecPointingUp()
+        vertical_dir = vertical_dir / np.linalg.norm(np.array(vertical_dir))
+        vertical_dir = V3D(vertical_dir[0], vertical_dir[1], vertical_dir[2])
+
+        peak_ip = peak_table.getPeak(0)
+        hkl_peak = np.array(peak_ip.getHKL())
+        hkl_peak = hkl_peak.reshape(-1, 1)
+        qsample_calc = np.array(np.pi * 2 * (ub_arb @ hkl_peak))
+        qsample_calc = V3D(float(qsample_calc[0]), float(qsample_calc[1]), float(qsample_calc[2]))
+        qsample_obs = np.array(peak_ip.getQSampleFrame())
+        qsample_obs = V3D(float(qsample_obs[0]), float(qsample_obs[1]), float(qsample_obs[2]))
+
+        if not np.degrees(qsample_calc.angle(vertical_dir)) < 5:
+            logger.warning("given peak hkl does not lie in the plane")
+
+        def rotation(theta):
+            axis = np.array(vertical_dir)
+            rotation_object = R.from_rotvec(theta * axis)
+            return rotation_object
+
+        def objective(theta):
+            rotation_object = rotation(theta)
+            rotated_vector = rotation_object.apply(qsample_calc)
+
+            cost = (np.linalg.norm(np.array(qsample_obs - rotated_vector))) ** 2
+            return cost
+
+        result = minimize(objective, 0, bounds=[(-np.pi, np.pi)], method="Nelder-Mead")
+        theta_new = result.x[0]
+        logger.debug(f"Angle of rotation applied to ub_arb {np.degrees(theta_new)}")
+
+        rotation_object = R.from_rotvec(theta_new * vertical_dir)
+        rotation_matrix = rotation_object.as_matrix()
+        ub_final = rotation_matrix @ ub_arb
+
+        self.exec_child_alg("SetUB", Workspace=peak_table, UB=ub_final)
+        logger.debug(f"Qs predicted for given hkl, {ub_final @ hkl_peak * 2 * np.pi}")
```

Comment:
As discussed offline, this is going to be moved to unit test assertions - it will true by construction here